modules/phase_field/src/materials/CrossTermBarrierFunctionMaterial.C

```diff
@@ -0,0 +1,73 @@
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "CrossTermBarrierFunctionMaterial.h"
+
+template<>
+InputParameters validParams<CrossTermBarrierFunctionMaterial>()
+{
+  InputParameters params = validParams<Material>();
+  params.addParam<std::string>("function_name", "g", "actual name for g(eta_i)");
+  MooseEnum h_order("SIMPLE=0", "SIMPLE");
+  params.addParam<MooseEnum>("g_order", h_order, "Polynomial order of the switching function h(eta)");
+  params.addRequiredCoupledVar("etas", "eta_i order parameters, one for each h");
+  params.addRequiredParam<std::vector<Real> >("W_ij", "Terms controlling barrier height set W=1 in DerivativeMultiPhaseMaterial for these to apply");
+  return params;
+}
+
+CrossTermBarrierFunctionMaterial::CrossTermBarrierFunctionMaterial(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Material>(parameters),
+    _function_name(getParam<std::string>("function_name")),
+    _g_order(getParam<MooseEnum>("g_order")),
+    _W_ij(getParam<std::vector<Real> >("W_ij")),
+    _num_eta(coupledComponents("etas")),
+    _eta(_num_eta),
+    _prop_g(declareProperty<Real>(_function_name)),
+    _prop_dg(_num_eta),
+    _prop_d2g(_num_eta)
+{
+  // if Vector W_ij is not the correct size to fill the matrix give error
+  if (_num_eta * _num_eta != _W_ij.size())
+    mooseError("Supply the number of etas squared for W_ij.");
+  // if W_ij is not symmetric or if diagonal values are not zero, return error
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    for (unsigned int j = 0; j < _num_eta; ++j)
+    {
+      if (_W_ij[_num_eta*i + j] != _W_ij[_num_eta*j + i])
+        mooseError("Supply symmetric values for W_ij");
+      if (_W_ij[_num_eta*i + i] != 0)
+        mooseError("Set on-diagonal values of W_ij to zero");
+    }
+
+  // declare derivative properties, fetch eta values
+  for (unsigned int i = 0; i < _num_eta; ++i)
+  {
+    const std::string & eta_name = getVar("etas", i)->name();
+    _prop_dg[i]  = &declarePropertyDerivative<Real>(_function_name, eta_name);
+    _prop_d2g[i] = &declarePropertyDerivative<Real>(_function_name, eta_name, eta_name);
+    _eta[i] = &coupledValue("etas", i);
+  }
+}
+
+void
+CrossTermBarrierFunctionMaterial::computeQpProperties()
+{
+
+  // Sum the components of our W_ij matrix to get constant used in our g function
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    for (unsigned int j = 0; j < _num_eta; ++j)
+    {
+
+      switch (_g_order)
+      {
+        case 0: // SIMPLE
+          _prop_g[_qp]         +=  _W_ij[_num_eta*i + j] * (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp];
+          (*_prop_dg[i])[_qp]  +=  2 * _W_ij[_num_eta*i + j] * 2 * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp];
```

Comment:
One 2* is meant for the derivative, the other is to count for the symmetric
other term in the W_ij matrix, which is why I put it in that form so I
would remember why. Because you said the prop_g counted each term twice but
the other one doesn't so I added the factor of two... I will switch it to a
4 though if you want.

On Thu, Jul 23, 2015 at 4:06 PM, Daniel Schwen <notifications@github.com>
wrote:

> In modules/phase_field/src/materials/CrossTermBarrierFunctionMaterial.C
> <https://github.com/idaholab/moose/pull/5437#discussion_r35377435>:
>
> > +}
> > +
> > +void
> > +CrossTermBarrierFunctionMaterial::computeQpProperties()
> > +{
> > +
> > +  // Sum the components of our W_ij matrix to get constant used in our g function
> > +  for (unsigned int i = 0; i < _num_eta; ++i)
> > +    for (unsigned int j = 0; j < _num_eta; ++j)
> > +    {
> > +
> > +      switch (_g_order)
> > +      {
> > +        case 0: // SIMPLE
> > +          _prop_g[_qp]         +=  _W_ij[_num_eta*i + j] * (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp];
> > +          (*_prop_dg[i])[_qp]  +=  2 * _W_ij[_num_eta*i + j] * 2 * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp];
>
> Wait a minute, is that right? shouldn't it be just one 2*?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5437/files#r35377435>.
>